File: trees/binaryTreeInorderTraversal.js
// https://leetcode.com/problems/binary-tree-inorder-traversal/
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// DFS
const inorderTraversal = (root) => {
  const stack = [];
  const res = [];

  while (root || stack.length) {
    if (root) {
      stack.push(root);
      root = root.left;
    } else {
      const node = stack.pop();
      res.push(node.val);
      root = node.right;
    }
  }

  return res;
};

// Visited flag
const inorderTraversal = (root) => {
  const stack = [[root, false]];
  const res = [];

  while (stack.length) {
    const [node, visitedFlag] = stack.pop();
    if (node) {
      if (visitedFlag) {
        res.push(node.val);
      } else {
        stack.push([node.right, false]);
        stack.push([node, true]);
        stack.push([node.left, false]);
      }
    }
  }

  return res;
};

File: trees/binaryTreeInorderTraversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # DFS
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        stack, res = [], []

        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                node = stack.pop()
                res.append(node.val)
                root = node.right

        return res


class Solution:
    # Visited flag
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        result, stack = [], [(root, False)]

        while stack:
            cur, visited = stack.pop()
            if cur:
                if visited:
                    result.append(cur.val)
                else:
                    stack.append((cur.right, False))
                    stack.append((cur, True))
                    stack.append((cur.left, False))

        return result


# recursively
def inorderTraversal1(self, root):
    res = []
    self.helper(root, res)
    return res


def helper(self, root, res):
    if root:
        self.helper(root.left, res)
        res.append(root.val)
        self.helper(root.right, res)
